core/shared/src/main/scala/scalaz/zio/ZManaged.scala

```diff
@@ -73,93 +213,516 @@ final case class ZManaged[-R, +E, +A](reserve: ZIO[R, E, Reservation[R, E, A]])
+
+  /**
+   * Executes this effect and returns its value, if it succeeds, but
+   * otherwise executes the specified effect.
+   */
+  final def orElse[R1 <: R, E2, A1 >: A](that: => ZManaged[R1, E2, A1]): ZManaged[R1, E2, A1] =
+    foldM(_ => that, ZManaged.succeed)
+
+  /**
+   * Returns an effect that will produce the value of this effect, unless it
+   * fails, in which case, it will produce the value of the specified effect.
+   */
+  final def orElseEither[R1 <: R, E2, B](that: => ZManaged[R1, E2, B]): ZManaged[R1, E2, Either[A, B]] =
+    foldM(_ => that.map(Right[A, B]), a => ZManaged.succeed(Left[A, B](a)))
+
+  /**
+   * Provides the `ZManaged` effect with its required environment, which eliminates
+   * its dependency on `R`.
+   */
+  final def provide(r: R): ZManaged[Any, E, A] =
+    provideSome(_ => r)
+
+  /**
+   * Provides some of the environment required to run this effect,
+   * leaving the remainder `R0`.
+   */
+  final def provideSome[R0](f: R0 => R): ZManaged[R0, E, A] =
+    Managed(reserve.provideSome(f).map(r => Reservation(r.acquire.provideSome(f), r.release.provideSome(f))))
+
+  /**
+   * Keeps some of the errors, and terminates the fiber with the rest.
+   */
+  final def refineOrDie[E1](pf: PartialFunction[E, E1])(implicit ev: E <:< Throwable): ZManaged[R, E1, A] =
+    refineOrDie(pf)(ev)
+
+  /**
+   * Keeps some of the errors, and terminates the fiber with the rest, using
+   * the specified function to convert the `E` into a `Throwable`.
+   */
+  final def refineOrDieWith[E1](pf: PartialFunction[E, E1])(f: E => Throwable): ZManaged[R, E1, A] =
+    catchAll { e =>
+      pf.lift(e).fold[ZManaged[R, E1, A]](ZManaged.die(f(e)))(ZManaged.fail)
+    }
+
+  /**
+   * Retries with the specified retry policy.
+   * Retries are done following the failure of the original `io` (up to a fixed maximum with
+   * `once` or `recurs` for example), so that that `io.retry(Schedule.once)` means
+   * "execute `io` and in case of failure, try again once".
+   */
+  final def retry[R1 <: R, E1 >: E, S](policy: ZSchedule[R1, E1, S]): ZManaged[R1 with Clock, E1, A] =
+    ZManaged {
+      self.reserve.map { case Reservation(acquire, release) => Reservation(acquire.retry(policy), release) }
+    }
+
+  final def right[R1 <: R, C]: ZManaged[Either[C, R1], E, Either[C, A]] = ZManaged.identity +++ self
+
+  /**
+   * Exposes the full cause of failure of this effect.
+   */
+  final def sandbox: ZManaged[R, Cause[E], A] =
+    ZManaged {
+      reserve.sandbox.map {
+        case Reservation(acquire, release) =>
+          Reservation(acquire.sandbox, release)
+      }
+    }
+
+  /**
+   * Zips this effect with its environment
+   */
+  final def second[R1 <: R, A1 >: A]: ZManaged[R1, E, (R1, A1)] = ZManaged.identity[R1] &&& self
+
+  /**
+   * Returns a new effect that executes this one and times the acquisition of the resource.
+   */
+  final def timed: ZManaged[R with Clock, E, (Duration, A)] =
```

Comment:
Makes sense. I'll update them

Reply:
I thought about this some more. Depending on how we end up implementing the *parN methods this can lead to unintuitive results if we also time the reservation